core/src/main/java/org/apache/iceberg/BaseIncrementalChangelogScan.java

```diff
@@ -133,51 +131,149 @@ private static Map<Long, Integer> computeSnapshotOrdinals(Deque<Snapshot> snapsh
+    private final Map<String, Integer> deleteFileToSnapshotOrdinal;
+
+    CreateDataFileChangeTasks(
+        long snapshotId,
+        Map<Long, Integer> snapshotOrdinals,
+        Map<String, Integer> deleteFileToSnapshotOrdinal) {
+      this.snapshotId = snapshotId;
+      this.snapshotOrdinals = snapshotOrdinals;
+      this.deleteFileToSnapshotOrdinal = deleteFileToSnapshotOrdinal;
+      this.changeOrdinal = this.snapshotOrdinals.get(snapshotId);
+    }
+
+    private DeleteFile[] filterAdded(DeleteFile[] deleteFiles) {
+      return FluentIterable.from(deleteFiles)
+          .filter(
+              deleteFile ->
+                  deleteFileToSnapshotOrdinal.get(deleteFile.path().toString()) == changeOrdinal)
+          .toArray(DeleteFile.class);
+    }
 
-    CreateDataFileChangeTasks(Deque<Snapshot> snapshots) {
-      this.snapshotOrdinals = computeSnapshotOrdinals(snapshots);
+    private DeleteFile[] filterExisting(DeleteFile[] deleteFiles) {
+      return FluentIterable.from(deleteFiles)
+          .filter(
+              deleteFile ->
+                  deleteFileToSnapshotOrdinal.get(deleteFile.path().toString()) < changeOrdinal)
+          .toArray(DeleteFile.class);
     }
 
     @Override
     public CloseableIterable<ChangelogScanTask> apply(
         CloseableIterable<ManifestEntry<DataFile>> entries, TaskContext context) {
 
-      return CloseableIterable.transform(
-          entries,
-          entry -> {
-            long commitSnapshotId = entry.snapshotId();
-            int changeOrdinal = snapshotOrdinals.get(commitSnapshotId);
-            DataFile dataFile = entry.file().copy(context.shouldKeepStats());
-
-            switch (entry.status()) {
-              case ADDED:
-                return new BaseAddedRowsScanTask(
-                    changeOrdinal,
-                    commitSnapshotId,
-                    dataFile,
-                    NO_DELETES,
-                    context.schemaAsString(),
-                    context.specAsString(),
-                    context.residuals());
-
-              case DELETED:
-                return new BaseDeletedDataFileScanTask(
-                    changeOrdinal,
-                    commitSnapshotId,
-                    dataFile,
-                    NO_DELETES,
-                    context.schemaAsString(),
-                    context.specAsString(),
-                    context.residuals());
-
-              default:
-                throw new IllegalArgumentException("Unexpected entry status: " + entry.status());
-            }
-          });
+      CloseableIterable<ChangelogScanTask> tasks =
+          CloseableIterable.transform(
+              entries,
+              entry -> {
+                long entrySnapshotId = entry.snapshotId();
+                DataFile dataFile = entry.file().copy(context.shouldKeepStats());
+                DeleteFile[] addedDeleteFiles = filterAdded(context.deletes().forEntry(entry));
```

Comment:
We can reuse `context.deletes().forEntry(entry)` later, it might worth to add a variable for it and reuse later

Reply:
You are right.